protocols/gossipsub/src/behaviour.rs

```diff
@@ -745,19 +744,25 @@ impl Gossipsub {
 
         // remove expired fanout topics
         {
-            let fanout = &mut self.fanout; // help the borrow checker
             let fanout_ttl = self.config.fanout_ttl;
-            self.fanout_last_pub.retain(|topic_hash, last_pub_time| {
-                if *last_pub_time + fanout_ttl < Instant::now() {
-                    debug!(
-                        "HEARTBEAT: Fanout topic removed due to timeout. Topic: {:?}",
-                        topic_hash
-                    );
-                    fanout.remove(&topic_hash);
-                    return false;
+            let now = Instant::now();
+            let expired_topics = self.fanout_last_pub.iter().filter_map(
```

Comment:
This two stage approach is necessary because BTreeMap does not yet have retain. It is in the works, but for now we first have to figure out what to remove, and then actually remove it.

For the probably quite common case where nothing is removed, this does not allocate anything.